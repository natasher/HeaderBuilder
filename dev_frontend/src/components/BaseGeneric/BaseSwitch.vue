<template>
  <div :class="classObject">

    <div class="switch" :data-name="fieldName">

      <span class = "path"   :data-name = "fieldName"></span>
      <span class = "handle" :data-name = "fieldName"></span>

    </div>
    <label>{{ label }}</label>

  </div>
</template>

<script>
export default {

  name: 'BaseSwitch',

  props: {
    /**
     * Label shown next to switch.
     */
    label: {
      type    : String,
      required: true,
    },
    /**
     * Determinates if switch is active,
     * takes function that returns `boolean` value
     * or boolean itself.
     */
    isActive: {
      type    : [ Function, Boolean ],
      required: false,
      default : false,
    },
    /**
     * The name of the containing/binded field.
     */
    fieldName: {
      type    : [ String, Boolean ],
      required: false,
      default : false,
    }
  },

  computed: {
    classObject: function () {
      return {
        'mfn-field mfnf-checkbox': true,
        'active'                 : this.isActive
      }
    }
  },

}
</script>
