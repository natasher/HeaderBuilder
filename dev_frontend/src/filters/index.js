/**
 * Capitalize first letter of str parameter.
 * e.g.
 *
 * "foo" => "Foo"
 * @param {string} str
 */
export const capitalize = ( value ) => {
  if ( !value ) return ''

  value = value.toString()
  return value.charAt( 0 ).toUpperCase() + value.slice( 1 )
}

/**
 * Convert string with dash separated words (e.g. value of data-type param)
 * to space separated string.
 * e.g.
 *
 * "menu_icon" => "menu icon"
 * @param {string} str
 */
export const convertDashToSpace = ( str ) => {
  if ( !str ) return ''

  return str.split( '_' ).join(' ')
}

/**
 * Convert string with space separated words to dash separated string.
 * e.g.
 *
 * "Action Bar" => "Action_Bar"
 * @param {string} str
 */
export const convertSpaceToDash = ( str ) => {
  if ( !str ) return ''

  return str.split(' ').join( '_' )
}

/**
 * Convert string with space to one Camel Cased string.
 * e.g.
 *
 * "background color" => "backgroundColor"
 * @param {string} str
 */
export const toCamelCase = ( str ) => {
  if ( !str ) return ''

  let arr    = str.split(' ')
  let output = arr.reduce(( prev, curr, index ) => {
    // return ( index === 0 ) ? prev + curr : prev + capitalize( curr )
    return (index === 0)
      ? curr.charAt(0).toLowerCase() + curr.slice(1)
      : prev + capitalize(curr)
  }, '')

  return output
}

/**
 * Convert camelCased string to human-readable string.
 * e.g.
 *
 * "backgroundColor" => "Background Color"
 * @param {string} str
 */
export const unCamelCase = ( str ) => {
  if ( !str ) return ''

  return str
    .replace(/([a-z])([A-Z])/g, '$1 $2')
    .replace(/\b([A-Z]+)([A-Z])([a-z])/, '$1 $2$3')
    .replace(/^./, ( str ) => { return str.toUpperCase() })
}

/**
 * Convert camelCased string to upper-cased snake case.
 * e.g.
 *
 * "secondRow" => "SECOND_ROW"
 */
export const allCaps = ( str ) => {
  if ( !str ) return ''

  return convertSpaceToDash(unCamelCase( str )).toUpperCase()
}