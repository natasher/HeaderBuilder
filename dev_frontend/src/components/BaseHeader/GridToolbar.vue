<template>
 <wrapper-clearfix class="general">

      <default-sticky-tabs class="tabs" />

      <wrapper-section title="Show rows"
        class="rows">
          <base-switch
            label         = "Action bar"
            :isActive     = "getActionBarActive"
            @click.native = "toggleActionBar" />

          <base-switch
            label         = "Second row"
            :isActive     = "getSecondRowActive"
            @click.native = "toggleSecondRow" />
      </wrapper-section>

      <wrapper-section title = "Layouts"
        class = "layouts"
        v-if = "getCurrentStoreModule === 'DesktopModule'">
        <layout-tabs class = "select-position" />
      </wrapper-section>

  </wrapper-clearfix>
</template>

<script>
import { mapState, mapGetters, mapActions } from 'vuex'

import WrapperClearfix   from '../BaseGeneric/WrapperClearfix.vue'
import WrapperSection    from '../BaseGeneric/WrapperSection.vue'
import BaseSwitch        from '../BaseGeneric/BaseSwitch.vue'
import DefaultStickyTabs from './DefaultStickyTabs.vue'
import LayoutTabs        from './LayoutTabs.vue'

export default {

  name: 'GridToolbar',

  components: {
    WrapperClearfix,
    WrapperSection,
    BaseSwitch,
    DefaultStickyTabs,
    LayoutTabs,
  },

  computed: {
    ...mapState( 'general', {
      getCurrentStoreModule: state => state.currentStoreModule,
    }),
    ...mapGetters( 'ui', [
      'getActionBarActive',
      'getSecondRowActive'
    ]),
  },

  methods: {
    ...mapActions( 'ui', [
      'toggleActionBar',
      'toggleSecondRow'
    ]),

  },
}
</script>
