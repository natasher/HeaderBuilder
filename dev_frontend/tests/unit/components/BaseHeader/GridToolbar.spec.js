import { shallowMount, createLocalVue } from '@vue/test-utils'
import { capitalize } from '../../../../src/filters/index'
import Vuex          from 'vuex'
import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'
import GridToolbar   from '../../../../src/components/BaseHeader/GridToolbar.vue'
import BaseSwitch    from '../../../../src/components/BaseGeneric/BaseSwitch.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      general: {
        namespaced: true,
        state: {
          currentStoreModule: 'DesktopModule',
        }
      },
      ui: {
        namespaced: true,
        getters: {
          getActionBarActive: () => { return false },
          getSecondRowActive: () => { return false },
        },
        actions: {
          toggleActionBar: jest.fn(),
          toggleSecondRow: jest.fn(),
        },
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    filters: { capitalize, },
    stubs: {
      BaseSwitch,
    }
  }

  return shallowMount( GridToolbar, merge( defaultMountingOptions, overrides ) )
}

describe('GridToolbar.vue', () => {

  describe('snapshots:', () => {

    test('should match if device is `desktop`', () => {
      const wrapper = createWrapper()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if device is NOT `desktop`', () => {
      const store = createStore({
        modules: {
          general: {
            state: {
              currentStoreModule: 'testModule'
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

  test('should invoke `toggleActionBar` when action bar switch is clicked', async () => {
    const toggleActionBar = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          actions: {
            toggleActionBar
          },
        }
      }
    })
    const wrapper = createWrapper({ store, })
    const actionbar_switch = wrapper.findAll(BaseSwitch).at(0)
    actionbar_switch.trigger('click')
    await flushPromises()

    expect( toggleActionBar ).toBeCalled()
  })

  test('should invoke `toggleSecondRow` when second row switch is clicked', async () => {
    const toggleSecondRow = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          actions: {
            toggleSecondRow
          },
        }
      }
    })
    const wrapper = createWrapper({ store, })
    const actionbar_switch = wrapper.findAll(BaseSwitch).at(1)
    actionbar_switch.trigger('click')
    await flushPromises()

    expect( toggleSecondRow ).toBeCalled()
  })

})