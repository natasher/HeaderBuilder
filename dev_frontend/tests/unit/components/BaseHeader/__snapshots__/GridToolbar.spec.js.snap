// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridToolbar.vue snapshots: should match if device is \`desktop\` 1`] = `
<wrapper-clearfix-stub
  class="general"
>
  <default-sticky-tabs-stub
    class="tabs"
  />
   
  <wrapper-section-stub
    class="rows"
    title="Show rows"
  >
    <div
      class="mfn-field mfnf-checkbox"
    >
      <div
        class="switch"
      >
        <span
          class="path"
        />
         
        <span
          class="handle"
        />
      </div>
       
      <label>
        Action bar
      </label>
    </div>
     
    <div
      class="mfn-field mfnf-checkbox"
    >
      <div
        class="switch"
      >
        <span
          class="path"
        />
         
        <span
          class="handle"
        />
      </div>
       
      <label>
        Second row
      </label>
    </div>
  </wrapper-section-stub>
   
  <wrapper-section-stub
    class="layouts"
    title="Layouts"
  >
    <layout-tabs-stub
      class="select-position"
    />
  </wrapper-section-stub>
</wrapper-clearfix-stub>
`;

exports[`GridToolbar.vue snapshots: should match if device is NOT \`desktop\` 1`] = `
<wrapper-clearfix-stub
  class="general"
>
  <default-sticky-tabs-stub
    class="tabs"
  />
   
  <wrapper-section-stub
    class="rows"
    title="Show rows"
  >
    <div
      class="mfn-field mfnf-checkbox"
    >
      <div
        class="switch"
      >
        <span
          class="path"
        />
         
        <span
          class="handle"
        />
      </div>
       
      <label>
        Action bar
      </label>
    </div>
     
    <div
      class="mfn-field mfnf-checkbox"
    >
      <div
        class="switch"
      >
        <span
          class="path"
        />
         
        <span
          class="handle"
        />
      </div>
       
      <label>
        Second row
      </label>
    </div>
  </wrapper-section-stub>
   
  <!---->
</wrapper-clearfix-stub>
`;
